Sources/Basics/HTTPClient.swift

```diff
@@ -81,43 +95,68 @@ public struct HTTPClient: HTTPClientProtocol {
         if request.options.addUserAgent, !request.headers.contains("User-Agent") {
             request.headers.add(name: "User-Agent", value: "SwiftPackageManager/\(SwiftVersion.currentVersion.displayString)")
         }
+        if let authorization = request.options.authorizationProvider?(request.url), !request.headers.contains("Authorization") {
+            request.headers.add(name: "Authorization", value: authorization)
+        }
         // execute
-        self._execute(request: request, requestNumber: 0) { result in
-            let callbackQueue = request.options.callbackQueue ?? self.configuration.callbackQueue
-            callbackQueue.async {
-                callback(result)
+        let callbackQueue = request.options.callbackQueue ?? self.configuration.callbackQueue
+        self._execute(
+            request: request, requestNumber: 0,
+            progress: progress.map { handler in
+                { received, expected in
+                    callbackQueue.async {
+                        handler(received, expected)
+                    }
+                }
+            },
+            completion: { result in
+                callbackQueue.async {
+                    completion(result)
+                }
             }
-        }
+        )
     }
 
-    private func _execute(request: Request, requestNumber: Int, callback: @escaping (Result<Response, Error>) -> Void) {
+    private func _execute(request: Request, requestNumber: Int, progress: ProgressHandler?, completion: @escaping CompletionHandler) {
         if self.shouldCircuitBreak(request: request) {
             diagnosticsEngine?.emit(warning: "Circuit breaker triggered for \(request.url)")
-            return callback(.failure(HTTPClientError.circuitBreakerTriggered))
+            return completion(.failure(HTTPClientError.circuitBreakerTriggered))
         }
 
-        self.underlying(request) { result in
-            switch result {
-            case .failure(let error):
-                callback(.failure(error))
-            case .success(let response):
-                // record host errors for circuit breaker
-                self.recordErrorIfNecessary(response: response, request: request)
-                // handle retry strategy
-                if let retryDelay = self.shouldRetry(response: response, request: request, requestNumber: requestNumber) {
-                    self.diagnosticsEngine?.emit(warning: "\(request.url) failed, retrying in \(retryDelay)")
-                    // TODO: dedicated retry queue?
-                    return self.configuration.callbackQueue.asyncAfter(deadline: .now() + retryDelay) {
-                        self._execute(request: request, requestNumber: requestNumber + 1, callback: callback)
+        self.underlying(
+            request,
+            { received, expected in
+                if let max = request.options.maximumResponseSizeInBytes {
```

Comment:
@yim-lee this is to better support "maximumResponseSizeInBytes" at the response buffering  time